CDS-92138 K8s Patch Step (#6481)
harness/developer-hub: docs/continuous-delivery/deploy-srv-diff-platforms/kubernetes/cd-k8s-ref/kubernetes-patch-step.md

---
title: Kubernetes Patch step
description: Make changes to specific workloads in a Kubernetes cluster without affecting other fields.
sidebar_position: 7
---

This topic describes the settings for the Kubernetes Patch step.

> The Patch step allows you to make changes to specific workloads in a Kubernetes cluster without affecting other fields. This step prevents downtime by making specific changes to the deployed service YAML. 

Reviewer: to make changes to specific resources/worklaods in

> The Patch step allows you to make changes to specific workloads in a Kubernetes cluster without affecting other fields. This step prevents downtime by making specific changes to the deployed service YAML. 

Author: ```suggestion
The Patch step allows you to make changes to specific resources or workloads in a Kubernetes cluster without affecting other fields. This step prevents downtime by making specific changes to the deployed service YAML. 
```

Here're some of the advantages of using a Patch step in your Kubernetes deployments:  

> - Efficiency: Instead of sending the entire resource specification for an update, you can send only the specific fields that need to be modified. This reduces the amount of data transferred over the network, and can improve the efficiency of resource updates, especially for large resources.

Author: ```suggestion
- Efficiency: Instead of sending the entire resource specification for an update, you can send the specific fields to be modified. This reduces the amount of data transferred over the network and improves the efficiency of resource updates, especially for large resources.
```

> - Granular control: The Patch command allows you to make precise changes to the Kubernetes resources by specifying which fields you want to modify. This level of granularity gives you more control over the update process and can help prevent unintended changes to other fields of the resource.

Author: ```suggestion
- Granular control: The Patch command allows you to make precise changes to the Kubernetes resources by specifying which fields you want to modify. This level of granularity gives you more control over the update process and helps prevent unintended changes to other fields of the resource.
```

> - Atomic updates: Kubernetes applies patches atomically, ensuring that either all changes are applied successfully or none of them are. This atomicity helps maintain consistency and integrity of resources in the cluster, especially when making multiple modifications simultaneously.

Reviewer 2: Instead of sending the entire resource specification for an update, you can send the specific fields to be modified. This reduces the amount of data transferred over the network and improves the efficiency of resource updates, especially for large resources.
- Granular control: The Patch command allows you to make precise changes to the Kubernetes resources by specifying which fields you want to modify. This level of granularity gives you more control over the update process and helps prevent unintended changes to other fields of the resource.

> - Concurrency: Since patch operations only modify specific fields of a resource, they can be performed concurrently with other operations on the same resource without the risk of conflicts or race conditions.

Reviewer 2: Kubernetes applies patches atomically, ensuring that either all changes are applied successfully or none of them are. This atomicity helps maintain the consistency and integrity of resources in the cluster, especially when making multiple modifications simultaneously.

> - Concurrency: Since patch operations only modify specific fields of a resource, they can be performed concurrently with other operations on the same resource without the risk of conflicts or race conditions.

Author: ```suggestion
- Concurrency: Kubernetes applies patches atomically, ensuring that either all changes are applied successfully or none of them are. This atomicity helps maintain the consistency and integrity of resources in the cluster, especially when making multiple modifications simultaneously.
```

- Compatibility: The patch command is compatible with various Kubernetes client libraries and API versions, allowing you to perform updates programmatically from different environments and programming languages.

## Prerequisites

- This feature is currently behind the feature flag, `CDS_K8S_PATCH_STEP_NG`. Contact [Harness support](mailto:support@harness.io) to enable it.
- This feature requires Harness Delegate version 828xx or later.
  
## Limitations

> - You will not be able to roll back your deployments with patched resources. Rolling back will only rollback your original deployment. 

Author: ```suggestion
- You will not be able to roll back your deployments with patched resources. Rollback will only roll back your original deployment. 
```

## Adding a Patch step

In your Kubernetes deployment pipeline's **Execution** tab, decide where you want to apply the patch and select **Add Step** > **Patch** to add a Patch step. 

## Step parameters

### Name

Name of the step.

### Workload

> Specify the workload to which you want to make changes. 

Reviewer: Specify the workload/resource to which you want to make changes, it needs to in format of "[namespace/]Kind/Name" similar to how we sepcify in Scale step

> Specify the workload to which you want to make changes. 

Author: ```suggestion
Specify the resource or workload to which you want to make changes in the format, `[namespace/]Kind/Name`. 
```

### Merge Strategy Type

> 

Reviewer 2: Merge strategy type

Author: its a UI element, step parameter

Reviewer 2: Sure please ignore the comment.

      
      In this example, let’s roll the nginx containers back to the latest image. This code tells `kubectl` to find the node at `/spec/template/spec/containers/0/image` and replace its value with the new image tag. That JSON path will take us to the image node in the first item in the containers list.

      
      ```json
      - op: replace
        path: "/spec/template/spec/containers/0/image"
        value: nginx:latest
      ```

      Here's the sample output after applying the JSON patch:  

      ```
      kubectl patch deployment nginx-deployment --type JSON --patch-file nginx_json_patch.yaml
      deployment.apps/nginx-deployment patched

      kubectl get pods -o jsonpath='{range .items[*]}{"\n"}{.metadata.name}{":\t"}{range .spec.containers[*]}{.image}{", "}{end}{end}' |grep nginx
      nginx-deployment-9b7747f45-4svqm:    nginx:latest, redis:latest,
      nginx-deployment-9b7747f45-hqjgh:    nginx:latest, redis:latest,
      ```
### Select Source

> 

Reviewer 2: Select source

Specify the source from where you want to access the patch file. Currently, we support Harness store, Git, and Inline sources. 

### Record Change Cause

> 

Reviewer 2: Record change cause

Select this option to record the patched operation including the patched content on the patched resource as the `kubernetes.io/change-cause` annotation. 

### Skip Steady State Check

> 

Reviewer 2: Skip steady state check
